[kan] Add new extractor

yhager opened the pull request:
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

New extractor for kan.org.il. 
Fixes #26551. 

This is my first extractor, I tried to follow the guide the best way I can, please let me know if there are any issues and I will address them.

dstftw commented on youtube_dl/extractor/kan.py:
```diff
@@ -0,0 +1,101 @@
+
+
+class KanIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?kan\.org\.il/(?:[iI]tem/\?item[iI]d|program/\?cat[iI]d)=(?P<id>[0-9]+)'
```
There must be two different extractors: for videos and for playlists.

dstftw commented on youtube_dl/extractor/kan.py:
```diff
@@ -0,0 +1,101 @@
+        creator = data.get('author', {}).get('name') or \
+            self._og_search_property('site_name', webpage, fatal=False)
+        thumbnail = get_thumbnail(data)
+        m3u8_url = data.get('content', {}).get('src')
```
This is mandatory. Read coding conventions.

dstftw commented on youtube_dl/extractor/kan.py:
```diff
@@ -0,0 +1,101 @@
+            video_id)
+        title = data.get('title') or \
+            self._og_search_title(webpage) or \
+            self._html_search_regex(r'<title>([^<]+)</title>', webpage, 'title')
```
This is never reachable.

dstftw commented on youtube_dl/extractor/kan.py:
```diff
@@ -0,0 +1,101 @@
+            self._html_search_regex(r'<title>([^<]+)</title>', webpage, 'title')
+        description = data.get('summary') or \
+            self._og_search_description(webpage, fatal=False)
+        creator = data.get('author', {}).get('name') or \
```
`try_get`.

dstftw commented on youtube_dl/extractor/kan.py:
```diff
@@ -0,0 +1,101 @@
+        m3u8_url = data.get('content', {}).get('src')
+        formats = self._extract_m3u8_formats(m3u8_url, video_id, ext='mp4')
+        return {
+            '_type': 'video',
```
This is default.

dstftw commented on youtube_dl/extractor/kan.py:
```diff
@@ -0,0 +1,101 @@
+
+    def _extract_list(self, list_id, webpage):
+        video_ids = re.findall(r'onclick="playVideo\(.*,\'([0-9]+)\'\)', webpage)
+        title = self._og_search_title(webpage)
```
Playlist title is optional.

dstftw commented on youtube_dl/extractor/kan.py:
```diff
@@ -0,0 +1,117 @@
+        creator = try_get(data, lambda x: x['author']['name'], str) or \
+            self._og_search_property('site_name', webpage, fatal=False)
+        thumbnail = get_thumbnail(data)
+        m3u8_url = try_get(data, lambda x: x['content']['src'], str)
```
Nothing changed.

dstftw commented on youtube_dl/extractor/kan.py:
```diff
@@ -0,0 +1,117 @@
+            self._og_search_property('site_name', webpage, fatal=False)
+        thumbnail = get_thumbnail(data)
+        m3u8_url = try_get(data, lambda x: x['content']['src'], str)
+        if not m3u8_url:
+            raise ExtractorError('Unable to extract m3u8 url')
```
Remove.

dstftw commented on youtube_dl/extractor/kan.py:
```diff
@@ -0,0 +1,117 @@
+            headers=self.geo_verification_headers())
+
+    def extract_item(self, video_id, webpage):
+        data = self._parse_json(
+            self._search_regex(
+                r'<script id="kan_app_search_data" type="application/json">([^<]+)</script>',
+                webpage,
+                'data',
+            ),
+            video_id,
+        )
```
Remove excessive verbosity. Read coding conventions.

dstftw commented on youtube_dl/extractor/kan.py:
```diff
@@ -0,0 +1,117 @@
+        title = data.get('title') or self._og_search_title(webpage)
+        description = data.get('summary') or \
+            self._og_search_description(webpage, fatal=False)
+        creator = try_get(data, lambda x: x['author']['name'], str) or \
```
compat_str

dstftw commented on youtube_dl/extractor/kan.py:
```diff
@@ -0,0 +1,117 @@
+            'id': video_id,
+            'title': title,
+            'thumbnail': thumbnail,
+            'formats': self._extract_m3u8_formats(m3u8_url, video_id, ext='mp4'),
```
This must be done right after title extraction.

dstftw commented on youtube_dl/extractor/kan.py:
```diff
@@ -0,0 +1,117 @@
+            'description': description,
+            'creator': creator,
+            'release_date': unified_strdate(data.get('published')),
+            'duration': parse_duration(data.get('extensions', {}).get('duration')),
```
try_get

dstftw commented on youtube_dl/extractor/kan.py:
```diff
@@ -0,0 +1,117 @@
+        video_ids = re.findall(r'onclick="playVideo\(.*,\'([0-9]+)\'\)', webpage)
+        entries = []
+        for video_id in video_ids:
+            url = 'https://www.kan.org.il/Item/?itemId=%s' % video_id
```
Do not shadow input `url`..

yhager commented:
Is there anything else I can do in order to get this extension merged?

aarubui commented:
For a start, follow "Trailing parentheses" section from the readme. Review `get_thumbnail` as it's defined outside the class and also only called from one place. `skip_download` for the tests needing `ffmpeg`. Unless it's just for me, geo bypass doesn't work so we can get rid of `_GEO_COUNTRIES` and `geo_verification_headers()`.

yhager commented:
@aarubui thanks for your review. I've fixed the trailing parens, added `skip_download` and removed `_GEO_COUNTRIES`. I'm not sure what you want to do with `get_thumbnail()`. Should I move this function to be in the class? or should I just unpack it and extract the thumbnail directly in the calling code?

aarubui commented:
- It's more common to use `self.playlist_result()` but I don't think you have to.
- `download_webpage` is now identical to `_download_webpage` and redundant.
- For `get_thumbnail`, I can see you need it as a method to break out of multiple loops. I've just noticed that, in other extractors, there are examples of functions defined both in and out of the classes. So you can ignore what I said and leave it as it is.
- "Trailing parentheses" wasn't for the import. It was fine before.
- If there's a risk of `<` appearing in the JSON, as part of the `summary` for example, you can try using `(.+?)` with `flags=re.DOTALL`.
- The site was geo-restricted to Israel yesterday but not today. You'll need [`skip`](https://github.com/ytdl-org/youtube-dl/blob/dfbbe2902fc67f0f93ee47a8077c148055c67a9b/youtube_dl/extractor/rai.py#L294) should that change back. You're fine for now.

Good luck with the real review.
